zookeeper/source/zk_dn_client.cc

```diff
@@ -0,0 +1,124 @@
+#ifndef RDFS_ZK_CLIENT_DN_CC
+#define RDFS_ZK_CLIENT_DN_CC
+
+#include "zk_dn_client.h"
+#include "zk_lock.h"
+#include <easylogging++.h>
+
+namespace zkclient{
+
+
+    const std::string ZkClientDn::CLASS_NAME = ": **ZkClientDn** : ";
+
+	ZkClientDn::ZkClientDn(const std::string& ip, const std::string& hostname, const std::string& zkIpAndAddress,
+            const uint32_t ipcPort, const uint32_t xferPort) : ZkClientCommon(zkIpAndAddress) {
+
+        data_node_id = DataNodeId();
+        data_node_id.ip = ip;
+        data_node_id.ipcPort = ipcPort;
+
+        // TODO: Fill in other data_node stats
+        data_node_payload = DataNodePayload();
+        data_node_payload.ipcPort = ipcPort;
+        data_node_payload.xferPort = xferPort;
+
+        registerDataNode();
+		LOG(INFO) << "Registered datanode " + build_datanode_id(data_node_id);
+
+	}
+
+	bool ZkClientDn::blockReceived(uint64_t uuid) {
+		// TODO: store file size
+		int error_code;
+		bool exists;
+
+		LOG(INFO) << "DataNode received a block with UUID " << std::to_string(uuid);
+        std::string id = build_datanode_id(data_node_id);
+
+		// Add acknowledgement
+		ZKLock queue_lock(*zk.get(), WORK_QUEUES + WAIT_FOR_ACK_BACKSLASH + std::to_string(uuid));
+		if (queue_lock.lock() != 0) {
+			LOG(ERROR) << CLASS_NAME <<  "Failed locking on /work_queues/wait_for_acks/<block_uuid> " << error_code;
+		}
+		if (zk->exists(WORK_QUEUES + WAIT_FOR_ACK_BACKSLASH + std::to_string(uuid), exists, error_code)) {
+			if (exists) {
+				if(zk->create(WORK_QUEUES + WAIT_FOR_ACK_BACKSLASH + std::to_string(uuid) + "/" + id, ZKWrapper::EMPTY_VECTOR, error_code, true)) {
+					LOG(ERROR) << CLASS_NAME <<  "Failed to create wait for acks " << error_code;
+				}
+			}
+		}
+		if (queue_lock.unlock() != 0) {
+			LOG(ERROR) << CLASS_NAME <<  "Failed unlocking on /work_queues/wait_for_acks/<block_uuid> " << error_code;
+		}
+
+		// Create a block_locations node for this block if there isn't one already
+		// TODO this should always exist
+		if (!zk->exists(BLOCK_LOCATIONS + std::to_string(uuid), exists, error_code)) {
```

Comment:
This line seems buggy. " !zk->exists " means that the exists lookup failed. The TODO is correct too though, the block location should always exist (although we have to careful with deletes). Barring strange race conditions, we should probably signal a FATAL or at least ERROR log or something if we have reached an inconsistent ZooKeeper state?